ethcore/src/client/client.rs

```diff
@@ -1551,23 +1556,35 @@ impl BlockChainClient for Client {
 	}
 
 	fn replay_block_transactions(&self, block: BlockId, analytics: CallAnalytics) -> Result<Box<Iterator<Item = Executed>>, CallError> {
-		let mut env_info = self.env_info(block).ok_or(CallError::StatePruned)?;
-		let body = self.block_body(block).ok_or(CallError::StatePruned)?;
-		let mut state = self.state_at_beginning(block).ok_or(CallError::StatePruned)?;
-		let txs = body.transactions();
-		let engine = self.engine.clone();
+		let block_hash = self.block_hash(block).ok_or(CallError::StatePruned)?;
+		let data_cached = {
+			self.replay_cache.lock().get_mut(&block_hash).is_some()
+		};
+		if data_cached {
+			let mut guard = self.replay_cache.lock();
+			let cached = guard.get_mut(&block_hash).unwrap();
+			Ok(Box::new(cached.clone().into_iter()))
+		} else {
+			let mut env_info = self.env_info(block).ok_or(CallError::StatePruned)?;
+			let body = self.block_body(block).ok_or(CallError::StatePruned)?;
+			let mut state = self.state_at_beginning(block).ok_or(CallError::StatePruned)?;
+			let txs = body.transactions();
+			let engine = self.engine.clone();
 
-		const PROOF: &'static str = "Transactions fetched from blockchain; blockchain transactions are valid; qed";
-		const EXECUTE_PROOF: &'static str = "Transaction replayed; qed";
+			const PROOF: &'static str = "Transactions fetched from blockchain; blockchain transactions are valid; qed";
+			const EXECUTE_PROOF: &'static str = "Transaction replayed; qed";
 
-		Ok(Box::new(txs.into_iter()
-			.map(move |t| {
-				let t = SignedTransaction::new(t).expect(PROOF);
-				let machine = engine.machine();
-				let x = Self::do_virtual_call(machine, &env_info, &mut state, &t, analytics).expect(EXECUTE_PROOF);
-				env_info.gas_used = env_info.gas_used + x.gas_used;
-				x
-			})))
+			let result: Vec<Executed> = txs.into_iter()
+				.map(move |t| {
+					let t = SignedTransaction::new(t).expect(PROOF);
+					let machine = engine.machine();
+					let x = Self::do_virtual_call(machine, &env_info, &mut state, &t, analytics).expect(EXECUTE_PROOF);
+					env_info.gas_used = env_info.gas_used + x.gas_used;
+					x
+				}).collect();
+			self.replay_cache.lock().insert(block_hash, result.clone());
```

Comment:
The cache has to take `Analytics` into consideration as well - currently you can get a cached result but for different `Analytics`.
Please try using `LruCache<(H256, Analytics), Vec<Results>>` as the cache key.